cql3/statements/select_statement.cc

```diff
@@ -434,8 +434,14 @@ select_statement::do_execute(query_processor& qp,
                         " you must either remove the ORDER BY or the IN and sort client side, or disable paging for this query");
     }
 
+    using maybe_result_msg = std::optional<shared_ptr<cql_transport::messages::result_message>>;
+
     if (_selection->is_trivial() && !_restrictions_need_filtering && !_per_partition_limit) {
-        return p->fetch_page_generator_result(page_size, now, timeout, _stats).then(wrap_result_to_error_message([this, p = std::move(p)] (result_generator&& generator) {
+      return repeat_until_value([this, page_size, now, timeout, p = std::move(p)] () mutable {
+        return p->fetch_page_generator_result(page_size, now, timeout, _stats).then(wrap_result_to_error_message([this, p = p.get()] (result_generator&& generator) -> maybe_result_msg {
+            if (!p->is_exhausted() && generator.empty()) {
+                return std::nullopt; // next page
+            }
```

Comment:
Shouldn't we return the empty page to the client?

Reply:
Turns out this code is not needed indeed. The python client (the one I'm using for tests) handles empty pages well.

Reply:
Although the python client can "handle" empty pages, I think if you want to test this code, you want to test individual pages, and see that the first page is empty. Otherwise how does the test test what you want it to test?

Reply:
The cql-python tests that I wrote check the query result page-by-page, making sure they have the expected number of empty pages and non-empty-pages.

Reply:
Oh, by the way, I forgot, but older clients indeed had bugs with empty pages - and I hve in test/cql-pytest/conftest a fixture "driver_bug_1" to check that the driver is new enough to support empty pages.
In any case, if you want to *check* that the page is empty, I think even older drivers will be fine. You just need to do a query and check that len(results.current_rows) == 0, you don't even need to continue the paging.

Reply:
Well, I do want to check that the row I expect is eventually delivered. You can check out my tests in the last commit.

Reply:
Good. I just now re-remembered how bad the cassandra-driver support for empty pages is. It was only fixed last year, but there was no new pip release of cassandra-driver in the last year and a half (!), so only scylla-driver also supports empty pages correctly.
That being said, I think it might be possible to write tests that work on the older driver by doing the paging manually and not letting the driver do it, but I think it's not worth the effort - the "driver_bug_1" fixture simply lets you check that the driver is new enough and if it isn't, skip the test.